pkg/controller/vsphere/session/tag_ids_caching_client.go

```diff
@@ -0,0 +1,194 @@
+	}
+	if item.(cachedItem).Expired() {
+		klog.V(4).Infof("cache item with name %s expired, invalidating", key)
+		c.Delete(key)
+		return "", false
+	}
+	return item.(cachedItem).value, true
+
+}
+
+func (c *cacheMap) Delete(key string) {
+	c.internalMap.Delete(key)
+}
+
+type tagsAndCategoriesCache struct {
+	tags       cacheMap
+	categories cacheMap
+}
+
+var sessionAnnotatedCache = map[string]*tagsAndCategoriesCache{}
+var sessionCacheMU sync.Mutex
+
+func getOrCreateSessionCache(sessionKey string) *tagsAndCategoriesCache {
+	sessionCacheMU.Lock()
+	defer sessionCacheMU.Unlock()
+
+	cache, found := sessionAnnotatedCache[sessionKey]
+	if !found {
+		cache = &tagsAndCategoriesCache{}
+		sessionAnnotatedCache[sessionKey] = cache
+		return cache
+	}
+	return cache
+}
+
+type CachingTagsManager struct {
+	*tags.Manager
+
+	sessionKey string // different vCenters might have a different tags in it
+}
+
+func newTagsCachingClient(tagsManager *tags.Manager, sessionKey string) *CachingTagsManager {
+	return &CachingTagsManager{
+		tagsManager, sessionKey,
+	}
+}
+
+// isName returns true if the id is not a urn.
+// this method came from vSphere sdk,
+// see https://github.com/vmware/govmomi/blob/a2fb82dc55a8eb00932233aa8028ce97140df784/vapi/tags/tags.go#L121 for
+// more context.
+func isName(id string) bool {
+	return !strings.HasPrefix(id, "urn:")
+}
+
+// isNotFoundErr checks if error message contains "Not Found" message.
+// vSphere api client does not expose error type, so we can rely only on error message
+func isNotFoundErr(err error) bool {
+	return err != nil && strings.HasSuffix(err.Error(), http.StatusText(http.StatusNotFound))
+}
+
+func (t *CachingTagsManager) GetTag(ctx context.Context, id string) (*tags.Tag, error) {
+	if !isName(id) { // if id is passed no cache check needed, use original GetTag method instantly
+		return t.Manager.GetTag(ctx, id)
+	}
+
+	cache := getOrCreateSessionCache(t.sessionKey)
+	cachedTagID, found := cache.tags.Get(id)
+	if found {
+		klog.V(4).Infof("found cached tag id value")
+		if cachedTagID == notFoundValue {
+			klog.V(4).Infof("cache contains special value indicates that tag was not found when cache was filled, treating as non existed tag")
+			return nil, fmt.Errorf(notFoundErrMessage)
+		}
+
+		tag, err := t.Manager.GetTag(ctx, cachedTagID)
+		if err != nil {
+			if strings.Contains(err.Error(), "404") {
+				klog.V(4).Infof("tag was not found in vCenter by cached id, invalidating cache")
+				// if tag not found, invalidate the cache and fallback to the default search method
+				cache.tags.Delete(t.sessionKey)
+				return t.Manager.GetTag(ctx, id)
+			}
+			return tag, err
+		}
+		return tag, nil
+	}
+
+	klog.V(4).Infof("tags cache miss, trying to find tag by name, it might take time")
+	tag, err := t.Manager.GetTag(ctx, id)
+	if err != nil {
+		if isNotFoundErr(err) {
+			klog.V(4).Infof("tag %s not found in vCenter, caching for 12 hours", id)
+			cache.tags.Set(id, notFoundValue)
```

Comment:
Gonna need some documentation to explain why we cache a not found, that could cause issues if users go later to create these